File: scraper_service/controllers/jobs.py
import pandas as pd
from datetime import datetime
from services.jobs import JobService
import os
from utils.source_loader import source_loader

class JobController:
    def __init__(self):
        self.sources = source_loader()  # Dynamically load all sources
        # self.sources= [Source1111()]

    def search_and_save_jobs(self, keyword):
        all_jobs = []
        total_count_all_sources = 0
        
        #TODO - 可以考慮用多執行緒的爬法
        for source in self.sources:
            print(f"正在處理資料來源: {source.__class__.__name__}")
            job_service = JobService(source)  # Create a JobService instance for each source
            total_count, jobs= job_service.search_jobs(keyword)
            
            all_jobs.extend(jobs)
            total_count_all_sources += total_count

        ## csv儲存
        df = pd.DataFrame(all_jobs)
        current_date_time = datetime.now()
        date_string = current_date_time.strftime("%Y%m%d")

        directory = f'data/records'

        if not os.path.exists(directory):
            os.makedirs(directory)

        csv_file_path = f'{directory}/jobs_{date_string}_{keyword}.csv'
        df.to_csv(csv_file_path, header=True, index=False)
        
        return total_count_all_sources, all_jobs

File: scraper_service_js/controllers/jobs.js
const path = require('path');
const { createObjectCsvWriter } = require('csv-writer');
const moment = require('moment');

const schedule = require('node-schedule');
const logger = require('../winston');
const { sourceLoader } = require('../utils/source_loader');
const JobService = require('../services/jobs');
const {
  JOB_STATUS, JOB_DURATION,
} = require('../metrics');

const searchAndSaveJobs = async (sources, keyword) => {
  let allJobs = [];
  let totalCountAllSources = 0;

  // 多腳本切換（可以優化效率）？不需要並行？
  for (const source of sources) {
    logger.info(`正在處理數據來源: ${source.constructor.name}`);
    const { totalCount, jobs } = await JobService.searchJobs(source, keyword);

    allJobs = allJobs.concat(jobs);
    totalCountAllSources += totalCount;
  }

  const csvWriter = createObjectCsvWriter({
    path: path.join('data', `jobs_${moment().format('YYYYMMDD')}_${keyword}.csv`),
    header: Object.keys(allJobs[0] || {}).map((key) => ({ id: key, title: key })),
  });
  await csvWriter.writeRecords(allJobs);
  return { totalCountAllSources, allJobs };
};

const job = async () => {
  try {
    const sources = await sourceLoader();
    if (sources.length > 0) {
      const keyword = 'node.js'; // 透過 API 修改關鍵字
      const end = JOB_DURATION.startTimer();
      try {
        await searchAndSaveJobs(sources, keyword);
      } finally {
        end(); // 計算 job 執行時間
      }
    }
  } catch (err) {
    logger.error('Error occurred:', { message: err.message, stack: err.stack });
  }
};

const scheduleJob = () => {
  try {
    schedule.scheduleJob('0 9 * * *', job);
    JOB_STATUS.set(1);
  } catch (err) {
    logger.error('Error occurred:', { message: err.message, stack: err.stack });
  }
};

module.exports = { job, scheduleJob };
